### PR TITLE
Using ImageJ.JS as the front-end for model test run

### PR DESCRIPTION
**Pull request by Nanguage**
image

### CODE REVIEW

**Comment by Nanguage** [comment]
TODO List:

- [x] Convert ImJoy plugin to a Vue component
- [x] Re shape the image to satisfy the spec
- [x] Change sample input to test-input(.npy file) by default
- [x] Add the code to wait imjoy start before imagej.js start
- [x] Fix the bug caused by image byte endian
- [x] Check if the item is a model or not
- [x] Check if the model is exist or not from: [the manifest](https://github.com/bioimage-io/bioengine-model-runner/blob/gh-pages/manifest.bioengine.yaml)
- [x] Check if the model is an img-to-img model, if it is, will show result using a table.
- [x] Create padder to allow user to input any xyz size image
- [x] Tiler for chunking the big images
- [x] Advance setting for set the tile size, input axes ...

File: src/components/ResourceItemInfo.vue [hunk 1]
```diff
@@ -204,7 +201,8 @@ export default {
     attachments: Attachments,
     "app-icons": AppIcons,
     "comment-box": CommentBox,
-    "resource-item-card": ResourceItemCard
+    "resource-item-card": ResourceItemCard,
+    TestRunForm
```
Comment (oeway) [inline src/components/ResourceItemInfo.vue]: 'test-run-form': TestRunForm

File: src/utils.js [hunk 2]
```diff
@@ -1065,3 +1065,52 @@ export function replaceAllRelByAbs(html, base_url) {
...
+      const end = [];
+      for (let j = 0; j < nDims; j++) {
+        const n = nTilesInEachDim[j];
+        start.push(
+          Math.floor(i / n) * (tileShape[j] - overlap[j]) +
+            (i % n) * tileShape[j]
+        );
+        end.push(start[j] + tileShape[j]);
+      }
+      tiles.push(new Tile(i, tileShape, start, end));
```
Comment (oeway) [inline src/utils.js]: Maybe it's better to use yield to create a generator here so we don't need to store all the tiles.

**Comment by Nanguage** [comment]
ezgif com-optimize

Now, by adding padding, it is possible to input images of any size. However, there are some models whose shape meets the spec but cannot work, such as the [affable-shark]() model. @oeway 

Details:

image

I think the shape of input conforms to the definition in the spec:

image

The [rdf](https://bioimage-io.github.io/collection-bioimage-io/rdfs/10.5281/zenodo.5764892/6647674/rdf.yaml) of shark model:

image

**Comment by oeway** [comment]
> Now, by adding padding, it is possible to input images of any size. However, there are some models whose shape meets the spec but cannot work, such as the affable-shark model. @oeway


Great! Thanks!
Will need to fix from the server side. Will get back to you once I fixed it.

Edit: Please also rebase from the main branch to allow merging the PR.

File: src/components/TestRun.vue [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+        throw err;
+      }
+      const rdf = ret.result.rdf;
+      console.log(rdf);
+      this.rdf = rdf;
+    },
+
+    async loadTritonClient() {
+      const server = await hyphaWebsocketClient.connectToServer({
+        server_url: "https://ai.imjoy.io",
```
Comment (oeway) [inline src/components/TestRun.vue]: Switch to https://hypha.bioimage.io

Comment (Nanguage) [reply src/components/TestRun.vue]: ok

File: src/components/TestRun.vue [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+      ]);
+      if (!resp.result.success) {
+        throw new Error(resp.result.error);
+      }
+      const outImg = resp.result.outputs[0];
+      return outImg;
+    },
+
+    async runModel() {
+      this.setInfoPanel("Running the model...", true);
```
Comment (oeway) [inline src/components/TestRun.vue:427]: The error handling should be better handled, we need to make sure that we capture all the error and keep the user informed with the interface. 

Currently one can stuck in "Running the model.." forever but there is an error like this:
```
Error: Based on the provided shape, [256,256,1], the tensor should have 65536 values but has 0
```

Ideally we should  do something like this:
```
this.setInfoPanel("Running the model...", true);
try{

    await this.runModel()
    this.setInfoPanel("Model run finished...")
}
catch(e){
    this.setInfoPanel("error....") // make sure we update the interface.
}

```

Comment (Nanguage) [reply src/components/TestRun.vue:427]: Okay, let me see how to change it

**Comment by Nanguage** [comment]
How to display the results of non-img-to-img models, such as the [HPA classification model](https://bioimage.io/#/?id=10.5281%2Fzenodo.5910854). One idea is to display it as a table. I'm not sure if ImageJ.JS can open a window that can embed HTML elements. Or we simply display the output array as an image. @oeway

**Comment by jmetz** [comment]
ImageJ should have a ResultsTable that tabular data can be displayed in. That would be very familiar for ImageJ users

**Comment by oeway** [comment]
> ImageJ should have a ResultsTable that tabular data can be displayed in. That would be very familiar for ImageJ users 

Good suggestion! Yes, we should use the result table, maybe via a imagej macro call.

Reference: https://forum.image.sc/t/creating-custom-tables-in-imagej/33100/2

**Comment by Nanguage** [comment]
### Show classification result as a table
bioimageio-classification-table

### Advance option allow user set the tile size and overlap
ezgif com-optimize

Could you give it a try? @oeway  If it's ok, I think it's time for merge.

**Review by oeway — APPROVED** [review]
Looks good!

**Pull request merged**